app/buck2_data/src/lib.rs

```diff
@@ -220,3 +221,27 @@ pub mod serialize_duration_as_micros {
         }
     }
 }
+
+/// Used to dispatch warnings to the event log in the case in the case of a
+/// retryable HTTP error. This often occurs when some level of exponential
+/// backoff is happening due to a lot of concurrent http_archive() requests for
+/// a single host; see GH-316 and GH-321 for background.
+pub trait DispatchableHttpRetryWarning {
+    /// Fire off a warning. This might go to the console, event log, or
+    /// some other place.
+    fn dispatch(&self, dur: &Duration, retries: usize, url: &str);
+}
+
+/// No-op implementation of DispatchableHttpRetryWarning. This does nothing at
+/// all when a retry occurs; useful for mock tests.
+pub struct NoopDispatchableHttpRetryWarning {}
+
+impl NoopDispatchableHttpRetryWarning {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl DispatchableHttpRetryWarning for NoopDispatchableHttpRetryWarning {
+    fn dispatch(&self, _backoff: &Duration, _retries: usize, _url: &str) {}
+}
```

Comment:
All of this should probably go into `buck2_common` instead of `buck2_data`? I merely put it there since I wrote it when I modified `data.proto`, so it was kind of coincidental.